### PR TITLE
Revert removing DCPU-16N

### PR DESCRIPTION
**Pull request by adam4813**
I am reverting the removal of the dcpu-16n, but I am not sure how "recent" changes would have affected things.

### CODE REVIEW

**Comment by Zardoz89** [comment]
Looking it, I don't see anything bad

**Comment by milesrout** [comment]
remind me, how did the DCPU-16N differ from the DCPU-16?

I'm actually currently working on a hardware implementation of the DCPU-16. Fun project, funny this arrived in my inbox only a week since I started working on it.

**Comment by Zardoz89** [comment]
The DCPU-16N extends the DCPU-16 with an Extended Memory Unit (EMU) to handle 24 bit addresses (16 MiB)

**Comment by Meisaka** [comment]
There are actually quite a few differences
- all hardware access is via either MMIO or via the _**I/O bus**_ (which is addressed a bit more like a flat memory).
- There are a few _system call-ish_ instructions, which in the opcode space, replace where the DCPU-16 would do hardware access
- The entire 16 bit address space goes through the "EMU" (a memory mapping system) to offer 24 bit bank switched memory access. This is fairly unrestricted, so if you swap out the page of an interrupt handler, bad things could happen.
- There are entire operational modes for accessing either 8 bit or 16 bit memories.
- There are few extra instructions to work with bytes easily

There are also a lot of similarities
- most of the core instruction set has the same meanings, mnemonics, and binary encodings as a DCPU-16
- the register file is the same
- while the DCPU-16 style hardware access isn't supported, 16N software can directly emulate it (although a bit slower)
- Interrupts work the same way
- The 16N is still an entirely unprotected execution model

for reference https://gist.github.com/Meisaka/8800367

File: tools/include/vm_parser.hpp [hunk 1]
```diff
@@ -178,6 +178,17 @@ struct VmParamaters {
...
+                    std::printf("Usage:\n");
+                    std::printf("\t%s -r romfile [other parameters]\n\n", argv[0]);
+                    std::printf("Parameters:\n");
+                    std::printf("\t-r file or --rom file : RAW binary file for the ROM 32 KiB\n");
+                    std::printf("\t-d file or --disk file : Disk file\n");
+                    std::printf("\t-c val or --cpu val : Sets the CPU to use, from \"tr3200\" or \"dcpu-16n\"\n");
+                    std::printf("\t-m val or --disk val : How many RAM have the computer in KiB. Must be > 128 and < 1024. Will be round to a multiple of 128\n");
+                    std::printf("\t--clock val : CPU clock speed in Khz. Must be 100, 250, 500 or 1000.\n");
+                    std::printf("\t-b val : Inserts a breakpoint at address val (could be hexadecimal or decimal).\n");
+                    std::printf("\t-h or --help : Shows this help\n");
```
Comment (Meisaka) [inline tools/include/vm_parser.hpp]: This should probably be removed, see the lines above it.

File: include/dcpu16n/dcpu16n.hpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * \file        dcpu16n.hpp
+ * \copyright   The MIT License (MIT)
+ *
+ */
+
+#ifndef __DCPU16N_HPP__
+#define __DCPU16N_HPP__
+
+#include "cpu.hpp"
+#include "vcomputer.hpp"
```
Comment (Meisaka) [inline include/dcpu16n/dcpu16n.hpp]: I think the path for these changed to "../*.hpp"

File: src/dcpu16n/dcpu16n.cpp [hunk 3]
```diff
@@ -0,0 +1,1047 @@
+/**
+* \brief       Class definitions for the DCPU-16N CPU
+* \file        dcpu16n.cpp
+* \copyright   The MIT License (MIT)
+*
+*/
```
Comment (Meisaka) [inline src/dcpu16n/dcpu16n.cpp]: The TR3200 specifies which version of the spec is implemented, this files does not, but should!
A bit of digging suggests that this implements DCPU-16N v0.8, however, 16N version 0.14 is the latest.
This also should link to the spec gist.
Unfortunately, this also means a good deal of this code is due for an overhaul.

File: src/dcpu16n/dis_dcpu16n.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
+/**
+ * Trillek Virtual Computer - DisDCPU16N.cpp
+ * OnLine dis-assembly of DCPU-16N machine code
+ */
```
Comment (Meisaka) [inline src/dcpu16n/dis_dcpu16n.cpp]: this handles v0.8 of the spec (old)

File: tools/include/vm_parser.hpp [hunk 5]
```diff
@@ -106,9 +106,9 @@ struct VmParamaters {
 
                     if (strncmp(arg, "tr3200", 6) == 0 || strncmp(arg, "TR3200", 6) == 0) {
                         cpu = CpuToUse::TR3200;
-                    }/* else if (strncmp(arg, "DCPU-16N", 6) == 0 || strncmp(arg, "dcpu-16n", 6) == 0) {
+                    } else if (strncmp(arg, "DCPU-16N", 6) == 0 || strncmp(arg, "dcpu-16n", 6) == 0) {
```
Comment (Meisaka) [inline tools/include/vm_parser.hpp]: "DCPU-16N" is a string of length 8, not 6. not that this is a very good way to check arguments mind you.

**Review by Meisaka — CHANGES_REQUESTED** [review]
There are few utility things that definitely need fixing.
Also, this would be version 0.8 of the DCPU-16N, it differs significantly from the current specs. This should either be explicitly noted, or updated.

File: tools/include/vm_parser.hpp (hunk 1 above)
Comment (adam4813) [reply tools/include/vm_parser.hpp]: 👍 this is the type of thing I wasn't certain about and that's why I wanted feedback!

File: tools/include/vm_parser.hpp (hunk 5 above)
Comment (adam4813) [reply tools/include/vm_parser.hpp]: Ha good call out

File: src/dcpu16n/dcpu16n.cpp [hunk 6]
```diff
@@ -0,0 +1,1047 @@
...
+    return vcomp->ReadW(0x00110000 | addr);
+}
+
+void DCPU16N::IOWrite(Word addr, Word v)
+{
+    vcomp->WriteW(0x00110000 | addr, v);
+}
+
+bool DCPU16N::DoesTrap(Word& msg) {
+	return false;
```
Comment (adam4813) [inline src/dcpu16n/dcpu16n.cpp:992]: I wasn't certain the expected behavior of DoesTrap, @Meisaka can you confirm what it should be?

Comment (Meisaka) [reply src/dcpu16n/dcpu16n.cpp:992]: from what I can tell, if this returns true, it will cause an interrupt to get sent back to the CPU towards the end of a "tick".
but since the 16N has a built-in queue, and my implementation used a crazy phase state machine, it's not really needed.
**tl;dr: just return false;**

**Review by Zardoz89 — APPROVED** [review]
I would leave to @Meisaka  criteria. I don't remember anything about this code base to say something useful.

File: include/dcpu16n/dcpu16n.hpp (hunk 2 above)
Comment (adam4813) [reply include/dcpu16n/dcpu16n.hpp]: Changed to relative

File: tools/include/vm_parser.hpp (hunk 1 above)
Comment (adam4813) [reply tools/include/vm_parser.hpp]: Removed duplicate help

File: src/dcpu16n/dcpu16n.cpp (hunk 3 above)
Comment (adam4813) [reply src/dcpu16n/dcpu16n.cpp]: I do not know where the relevant gist is located, do you have the link?

File: tools/include/vm_parser.hpp (hunk 5 above)
Comment (adam4813) [reply tools/include/vm_parser.hpp]: Changed to 8 for the time.

File: src/dcpu16n/dcpu16n.cpp (hunk 3 above)
Comment (adam4813) [reply src/dcpu16n/dcpu16n.cpp]: Found it https://gist.github.com/Meisaka/8800367

**Review by Meisaka — APPROVED** [review]
It at least looks merge-able to me, so as long as it compiles I would say it's good
